### ISSUE TITLE
C:\python3\python.exe does not exist while running pre-commit on Windows

### ISSUE THREAD

**Issue by dawelter2** [pre-commit/pre-commit#913]
Hi,

I'm trying to run the pre-commit commands of the [ESSS/barril](https://github.com/ESSS/barril) repository on my Windows machine, but I'm getting this error:

```
 λ tox -e linting
linting installed: aspy.yaml==1.1.1,cfgv==1.4.0,identify==1.1.8,importlib-metadata==0.8,importlib-resources==1.0.2,nodeenv==1.3.3,pre-commit==1.14.0,PyYAML==3.13,six==1.12.0,toml==0.10.0,virtualenv==16.2.0,zipp==0.3.3
linting run-test-pre: PYTHONHASHSEED='554'
linting runtests: commands[0] | pre-commit run --all-files --show-diff-on-failure
[INFO] Installing environment for https://github.com/ambv/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: Command: ('w:\\projects\\barril\\.tox\\linting\\scripts\\python.exe', '-mvirtualenv', 'C:\\Users\\darci\\.cache\\pre-commit\\repoufam_s8_\\py_env-python3', '-p', 'C:\\python3\\python.exe')
Return code: 3
Expected return code: 0
Output:
    The path C:\python3\python.exe (from --python=C:\python3\python.exe) does not exist

Errors: (none)

Check the log at C:\Users\darci/.cache\pre-commit\pre-commit.log
```

The `C:\python3` path does not exist on my machine, I'm using `conda` and `virtualenv` to create a Python 3.6 virtual environment, from which I execute `tox -e linting`.

Here's the contents of the `pre-commit-config.yaml`:

```yaml
repos:
-   repo: https://github.com/ambv/black
    rev: 18.6b4
    hooks:
    -   id: black
        args: [--safe, --quiet]
        language_version: python3
-   repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v1.3.0
    hooks:
    -   id: trailing-whitespace
    -   id: end-of-file-fixer
    -   id: check-yaml
    -   id: debug-statements
    -   id: flake8
-   repo: local
    hooks:
    -   id: rst
        name: rst
        entry: rst-lint --encoding utf-8
        files: ^(CHANGELOG.rst|HOWTORELEASE.rst|README.rst)$
        language: python
        additional_dependencies: [pygments, restructuredtext_lint]
```

Any hints of what the problem might be?

cc @nicoddemus

**Comment by asottile** [comment]
I'm not sure how conda sets things up but standard python installations on windows are detected via three ways (quite similar to tox, but perhaps can be extended to do this [one thing](https://github.com/tox-dev/tox/blob/554548434d2b0a84f0e8c1d74d221a97cf5e03b8/src/tox/interpreters.py#L143-L144) tox does):

1. look up as if it is a posix executable, that is look for `python3.exe` somewhere
2. look up using the `py` launcher, the standard way to register executables on windows (`py -3` in this case)
3. all else fails, assume the "standard" installation location of `C:\PythonXX\python.exe` -- this is the final one that's shown in the error message

My guess is that conda does not participate in any of those three approaches and perhaps the approach from tox should be pursued.  

If you could provide the output of `py -l` and `which python3` I could probably confirm or deny my suspicions :)

**Comment by asottile** [comment]
can you try the branch in #914 -- it might fix this though I don't have a conda setup to test

**Comment by nicoddemus** [comment]
Hi @asottile,

@dawelter2 is my colleague, in fact I wrote the original post from his computer. 😁 

You are right, none of those 3 methods work in a computer with just `conda` installed.

`tox` works fine in this setup though, so perhaps extending `pre-commit` to do the same might fix this. Although, looking at the code:

```python
if envconfig.basepython == "python{}.{}".format(*sys.version_info[0:2]):
    return sys.executable
```

This won't work if a user specifies `basepython = python3` because it will try to check against `python3.6` in a Python 3.6 installation. Perhaps adding a new *last* step:

```python
if envconfig.basepython == "python{}".format(*sys.version_info[0]):
    return sys.executable
```

Would help in that case?

**Comment by asottile** [comment]
yeah the tox code is non-ideal, I wrote a better version with some tests: https://github.com/pre-commit/pre-commit/pull/914/files#diff-235f2434ddb56aaeef64d44558e82943R92

Now that I think about it, I was hitting this in travis-ci windows 2.7 and just checked and it fixed my build there: https://travis-ci.org/asottile-archive/pyupgrade/jobs/477862429

so I think/hope this should fix this issue as well 😆

**Comment by nicoddemus** [comment]
@asottile weird, we get the same error... looking at your code it should work. We are using `python 3.6.5` here.

**Comment by asottile** [comment]
Hmmm i wish I could poke around with the debugger, mind running those functions and pasting their return values?

**Comment by nicoddemus** [comment]
```
```
import sys
def _sys_executable_matches(version):
    if version == 'python':
        return True
    elif not version.startswith('python'):
        return False

    try:
        info = tuple(int(p) for p in version[len('python'):].split('.'))
    except ValueError:
        return False

    return sys.version_info[:len(info)] == info
	
	
print(_sys_executable_matches('python3'))
print(_sys_executable_matches('python3.6'))
```

```
(.env) λ python --version
Python 3.6.5 :: Anaconda, Inc.

(.env) λ python foo.py
True
True
```

😞

**Comment by asottile** [comment]
Hmmm, how about this function : https://github.com/pre-commit/pre-commit/blob/v1.14.0/pre_commit/languages/python.py#L46

**Comment by nicoddemus** [comment]
```
```
(.env) λ python                                                                                     
Python 3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)] on win32 
Type "help", "copyright", "credits" or "license" for more information.                              
>>> from pre_commit.languages.python import _get_default_version                                    
>>> print(_get_default_version())                                                                   
default                                                                                             
>>>         
```

**Comment by asottile** [comment]
alright let me get a windows vm :(

**Comment by asottile** [comment]
I was able to reproduce the original problem using conda

I was then able to fix it by running:

```
conda uninstall pre_commit
pip install git+https://github.com/pre-commit/pre-commit@self_exe_python
```

I'll release 1.14.1 later with the fix

**Comment by asottile** [comment]
This is available in [v1.14.1](https://github.com/pre-commit/pre-commit/releases/tag/v1.14.1) -- thanks again for the issue @dawelter2  and the assist @nicoddemus :tada:

**Comment by nicoddemus** [comment]
Awesome, thanks for the quick response! 😁
